### ISSUE TITLE
download.webmin.com present bad and expired certificates

### ISSUE THREAD

**Issue by Sn4kY** [webmin/webmin#1533]
Hi Jamie

the 104.207.151.13.vultr.com  server, which provides download.webmin.com still sent ISRG Root X1 cross signed with expired DST Root CA X3
expand path in SSLLabs 
https://www.ssllabs.com/ssltest/analyze.html?d=download.webmin.com&s=104.207.151.13&latest
this one is correct
https://www.ssllabs.com/ssltest/analyze.html?d=download.webmin.com&s=108.60.199.109&latest

Could you correct this ? Debian 7 and 8 are unable to connect to https://download.webmin.com (many thanks to have leaved http:// without redirect to https)
Cheers

**Comment by swelljoe** [comment]
>The solution is to force a certificate renewal server-side, which will then not include the expired certification path anymore. It may however be possible that older distros/clients will still not connect, if their CA store does not contain the new ISRG Root X1 root certificate yet.

That's not accurate. By default, new certificates are [still cross-signed by that expired cert, intentionally](https://letsencrypt.org/2020/12/21/extending-android-compatibility.html). You have to use a very new certbot version (1.19, newer than almost any distro has) and explicitly request the alternate intermediate chain using the `--preferred-chain` option (https://certbot.eff.org/docs/using.html#certbot-command-line-options).

I assumed new certs would not be cross-signed with the expired cert, as well, but I did some testing and confirmed they're still signing certs with it, in the default configuration. One can argue about whether they're right to do so, but that's what they're doing.

Now, whether download.webmin.com should be signed with the cross-signed version is probably a useful conversation to have...but just regenerating the cert won't change it.

**Comment by swelljoe** [comment]
Also, I'll mention there is a client-side solution that works with the old cross-signed cert.

If you edit `/etc/ca-certificates.conf` and add a `!` before the offending cert, and then run `update-ca-certificates`, it'll resolve the problem locally. The problem is a buggy openssl version...not the actual cert. There _is_ a valid certification path for these certs...the client is just failing to find it due to that bug.

e.g. in the ca-certificates.conf file make it look like this:

```
!mozilla/DST_Root_CA_X3.crt
```

**Comment by swelljoe** [comment]
I only know all this because I fought with it on my own fleet of devices all weekend and was baffled by the behavior, and by the decisions of the LE folks who seemed to be intentionally shipping broken certs. (I still kinda question their judgement here, as it seems like breaking all older Ubuntu and Debian versions is worse than breaking very old Android versions, but they have more data than I do and are much smarter than I am, at least about anything SSL/TLS related).

**Comment by MichaIng** [comment]
> it seems like breaking all older Ubuntu and Debian versions is worse

which are broken regardless whether the expired path is shipped additionally, because... it expired 😄.

Thanks for the clarification. The Debian System Administrators managed to do it for `deb.debian.org`/`cdn-fastly.deb.debian.org`, but I didn't know about the possibly required steps with Certbot. I mean there are other ACME clients, if LE was used, probably other clients do it right OOTB. I just checked back with my own server, just renewed via `acme.sh`, and the fullchain contains only the three intended certs of the new path. SSL Labs currently still shows the expired one as well, but probably a caching issue.

**Comment by swelljoe** [comment]
No, it's not broken because the cross-sign is expired, it's broken because of an openssl bug. The current R3 certs have a working validation path, but the buggy openssl version doesn't find it.

**Comment by swelljoe** [comment]
> I just checked back with my own server, just renewed via acme.sh, and the fullchain contains only the three intended certs of the new path. SSL Labs currently still shows the expired one as well, but probably a caching issue.

Nope. SSL Labs is right. You're not reading the fullchain right. The cross-sign cert is not included in the fullchain.pem. You won't see it. You'll see the R3 intermediate, which is cross-signed by the expired DST cert. 

Again, I only know all this because my brain went through all the same confusions you're going through, and it took me a long time to grok, so I'm trying to save you some time/confusion.

**Comment by jcameron** [comment]
@swelljoe can or should we fix this by including another cert in the chain file for download.webmin.com ?

**Comment by swelljoe** [comment]
I linked to Let's Encrypt's explanation for why they're doing what they're doing in continuing to cross-sign with the expired certificate. 

> I'm not talking about EOL clients here, which do not have ISRG Root X1 in their root CA store

Nor am I.

I'm talking about OpenSSL choosing an invalid validation path when a valid path is available. That is, according the Let's Encrypt folks, a bug. I'm not an expert on OpenSSL, but I assume they are.
